neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsList.java

```diff
@@ -0,0 +1,285 @@
+            }
+        }
+
+        throw new IndexOutOfBoundsException(String.format("index=%s, size=%s", index, size));
+    }
+
+    @Nonnull
+    @Override
+    public ListIterator<E> listIterator(int index) {
+        checkPositionIndex(index, features.size());
+
+        ListIterator<E> iterator = features.isEmpty()
+                ? ContentsListIterator.empty()
+                : new ContentsListIterator<>(owner, features, resolve);
+
+        // Advance the cursor the desired index
+        for (int i = 0; i < index; i++) {
+            iterator.next();
+        }
+
+        return iterator;
+    }
+
+    @Nonnegative
+    @Override
+    public int size() {
+        int result = 0;
+
+        for (EStructuralFeature feature : features) {
+            final Object value = owner.eGet(feature, false);
+
+            // Feature is feature map
+            if (FeatureMapUtil.isFeatureMap(feature)) {
+                for (FeatureMap.Entry e : FeatureMap.class.cast(value)) {
+                    if (isIncludedEntry(e)) {
+                        result++;
+                    }
+                }
+            }
+
+            // Feature is multi-valued
+            else if (feature.isMany()) {
+                result += Collection.class.cast(value).size();
+            }
+
+            // Feature is single-valued
+            else if (nonNull(value)) {
+                result++;
+            }
+        }
+
+        return result;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        for (EStructuralFeature feature : features) {
+            final Object value = owner.eGet(feature, false);
+
+            // Feature is feature map
+            if (FeatureMapUtil.isFeatureMap(feature)) {
+                for (FeatureMap.Entry e : FeatureMap.class.cast(value)) {
+                    if (isIncludedEntry(e)) {
+                        return false;
+                    }
+                }
+            }
+
+            // Feature is multi-valued
+            else if (feature.isMany()) {
+                if (!Collection.class.cast(value).isEmpty()) {
+                    return false;
+                }
+            }
+
+            // Feature is single-valued
+            else if (nonNull(value)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Nonnull
+    @Override
+    public List<E> basicList() {
+        return nonResolvingList.get();
+    }
+
+    @Override
+    public void move(int newPosition, E e) {
+        throw new UnsupportedOperationException("move");
+    }
+
+    @Override
+    public E move(int newPosition, int oldPosition) {
+        throw new UnsupportedOperationException("move");
+    }
+
+    /**
+     * An empty {@code ContentsList}.
+     *
+     * @param <E> the type of elements in this list
+     */
+    @ParametersAreNullableByDefault
+    private static class EmptyContentsList<E extends EObject> extends ContentsList<E> {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This class has 6 parents which is greater than 5 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)